enhancements/node-tuning/mixed-cpu-node-plugin.md

```diff
@@ -0,0 +1,352 @@
+
+### Workflow Description
+
+The premise is that OCP cluster is running, and there's an active performance-profile that already tuned the system.
+
+1. The cluster administrator wants to support both exclusive and shared cpus for workloads running on the cluster.
+2. The cluster administrator specifies the shared CPU ids in the PerformanceProfile `spec.cpu.shared`.
+3. The cluster administrator waits for MCO to kick in, update and reboot the nodes.
+4. The cluster administrator waits for the node to come back from reboot.
+5. The application administrator wants to deploy their DPDK application as a Guaranteed pod with shared CPUs.
+6. The application administrator specifies a request for `openshift.io/shared-cpus: 1` under the pod's `spec.containers[].resources.requests`.
+7. The application administrator is waiting for the DPDK pod to be `Running`.
+8. The DPDK's app user/developer wants to run a light-weight task (threads) on shared cpus.
+9. The DPDK's app user/developer should pin the light-weight threads to CPUs that have shown in the `OPENSHIFT_SHARED_CPUS` environment variable of the container's process.
+
+### API Extensions
+N/A
+
+### Risks and Mitigations
+
+* With this solution, the platform's housekeeping processes also runs on the shared CPUs.
+but the intent is to allocate those cpus to workload’s light-weight tasks,
+so having some latency is bearable.
+A way of mitigating that is checks whether it's possible to use workload partitioning
+to ensure the platform housekeeping processes don't run on the shared cpus.
+(This statement needs to be verified)
+
+* There is no risk that workload’s tasks run on dedicated OCP’s housekeeping reserved cpus,
+because only the shared cpus exposed via cgroups to the workload’s process.
+In other words, platform's housekeeping processes can run on shared cpus dedicated to workload’s light-weight tasks,
+but not the other way around.
+
+* mixed-cpus-node-plugin is NRI-based plugin.
+Bugs in the plugin or NRI framework might affect the containers at various lifecycle stages, e.g.,
+CreateContainer, UpdateContainer, etc.
+For mitigating such cases, it is possible to deactivate the node-plugin, and in severe cases it is possible to disable
+NRI completely at runtime-level.
+
+* From a performance perspective, the runtime communicates with the node-plugin via gRPC calls.
+Therefore, for every container creation, there's another gRPC call between the runtime and the node-plugin.
+We can't avoid this call even for containers that don't ask for shared CPUs, since runtime doesn't have any filtering
+mechanism, it sends all the containers through the NRI framework to node-plugin, and the filtering is done at the
+plugin side.
+
+### Drawbacks
+
+The way of how user should ask for shared CPUs in the pod spec is through device plugin.
+While this approach fits in the model of how the workload requires resources,
+better integration would require more invasive changes which are out of scope now.
```

Comment:
can you elaborate on what a better integration would be and why just to understand the eventual desired state?

Reply:
A better integration would be asking for shared cpus directly as part of the pod spec requests and not via a device request api as we are doing now. 
That would be cleaner and better user experience.

So basically instead of having:
```yaml
kind: pod
spec:
 containers:
 - resources:
    requests:
     memory: "100Mi"
     cpu: "2"
     openshift.io/shared-cpu: 1
```
Which is essentially a device request.

We would like to something similar to this:
```yaml
kind: pod
spec:
 containers:
 - resources:
    requests:
     memory: "100Mi"
     cpu: "2"
     shared-cpu: "1"
```

But the existing k8s pod spec architecture doesn't allow you to have a request for both exclusive and shared cpus for the same container and explicitly quantify each.
Integration of such significant change in the pod spec API is something that we're no aiming for in this EP.